test/e2e/prometheus_test.go

```diff
@@ -0,0 +1,312 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const monitoringNamespace string = "monitoring"
+
+// Agent metrics to validate
+var antreaAgentMetrics = []string{
+	"antrea_agent_local_pod_count",
+	"antrea_agent_ovs_flow_table",
+	"antrea_agent_runtime_info",
+}
+
+// Controller metrics to validate
+var antreaControllerMetrics = []string{
+	"antrea_controller_address_group_processed",
+	"antrea_controller_address_group_sync_duration_milliseconds",
+	"antrea_controller_applied_to_group_processed",
+	"antrea_controller_applied_to_group_sync_duration_milliseconds",
+	"antrea_controller_length_address_group_queue",
+	"antrea_controller_length_applied_to_group_queue",
+	"antrea_controller_length_network_policy_queue",
+	"antrea_controller_network_policy_processed",
+	"antrea_controller_network_policy_sync_duration_milliseconds",
+	"antrea_controller_runtime_info",
+}
+
+var prometheusEnabled bool
+
+// Prometheus server JSON output
+type prometheusServerOutput struct {
+	Status string
+	Data   []map[string]string
+}
+
+func init() {
+	flag.BoolVar(&prometheusEnabled, "prometheus", false, "Enables Prometheus tests")
+}
+
+// skipIfPrometheusDisabled checks if Prometheus testing enabled, skip otherwise
+func skipIfPrometheusDisabled(t *testing.T) {
+	if !prometheusEnabled {
+		t.Skip("Prometheus testing is disabled")
+	}
+}
+
+// getMonitoringAuthToken retrieves monitoring authorization token, required for access to Antrea apiserver/metrics
+// resource
+func getMonitoringAuthToken(t *testing.T, data *TestData) string {
+	secrets, err := data.clientset.CoreV1().Secrets(monitoringNamespace).List(metav1.ListOptions{})
+	if err != nil {
+		t.Fatalf("Error fetching monitoring secrets: %v", err)
+	}
+
+	var token string
+	for _, secret := range secrets.Items {
+		if strings.HasPrefix(secret.Name, "prometheus-token") {
+			token = string(secret.Data[v1.ServiceAccountTokenKey])
+		}
+	}
+
+	if len(token) == 0 {
+		t.Fatal("Prometheus ServiceAccount secret not found")
+	}
+
+	return token
+}
+
+// getMetricsFromApiServer retrieves Antrea metrics from Pod apiserver
+func getMetricsFromApiServer(t *testing.T, url string, token string) string {
+	config := &tls.Config{
+		InsecureSkipVerify: true,
+	}
+	tr := &http.Transport{TLSClientConfig: config}
+	client := &http.Client{Transport: tr}
+
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		t.Fatalf("Error creating HTTP request: %v", err)
+	}
+
+	if token != "" {
+		req.Header.Add("Authorization", "Bearer "+token)
+	}
+
+	// Query metrics via HTTPS from Pod
+	resp, err := client.Do(req)
+	if err != nil {
+		t.Fatalf("Error fetching metrics from %s: %v", url, err)
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		t.Fatalf("Error retrieving metrics from %s. response: %v", url, err)
+	}
+
+	return string(body)
+}
+
+// testPrometheusMetricsOnPods locate Antrea Pods from the specified component, retrieve and validate that supplied
+// metric exists
```

Comment:
testPrometheusMetricsOnPods locates Antrea Pods from the specified component, then retrieves and validates that all the supplied metrics exist